src/Microsoft.VisualStudio.Services.Agent/ProcessInvoker.cs

```diff
@@ -288,6 +294,235 @@ private void StartReadStream(StreamReader reader, ConcurrentQueue<string> dataBu
                 }
             });
         }
+
+#if OS_WINDOWS
+        private async Task WindowsCancelAndKillProcessTree()
+        {
+            bool ctrl_c_succeed = await SendCtrlSignal(ConsoleCtrlEvent.CTRL_C, TimeSpan.FromSeconds(10));
+            if (ctrl_c_succeed)
+            {
+                Trace.Info("Process cancelled successfully through Ctrl+C.");
+                return;
+            }
+
+            bool ctrl_break_succeed = await SendCtrlSignal(ConsoleCtrlEvent.CTRL_BREAK, TimeSpan.FromSeconds(5));
+            if (ctrl_break_succeed)
+            {
+                Trace.Info("Process terminate successfully through Ctrl+Break.");
+                return;
+            }
+
+            Trace.Info("Kill entire process tree since both cancel and terminate signal has been ignored by the target process.");
+            KillProcessTree(_proc.Id);
+        }
+
+        private async Task<bool> SendCtrlSignal(ConsoleCtrlEvent signal, TimeSpan timeout)
+        {
+            Trace.Info($"Sending {signal.ToString()} to process {_proc.Id}.");
+            ConsoleCtrlDelegate ctrlEventHandler = new ConsoleCtrlDelegate(ConsoleCtrlHandler);
+            try
+            {
+                if (!FreeConsole())
```

Comment:
Do we run the worker.exe directly from any tests? If so would this interfere with those tests? *I think* running the worker is a goal for L1's (future). It may be worth leaving a big comment here explaining the implication.